### PR TITLE
Task05

### LINKED ISSUE #5: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [x] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [x] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [x] Where in the code should we implement this functionality?
- [x] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by lei-appfolio**
Closes #5 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/views/images/index.html.erb [hunk 1]
```diff
@@ -26,6 +30,12 @@
   <%= link_to 'Submit Image', new_image_path %>
 </p>
 
+<% unless @search_tag.blank? %>
+  <p>
```
Comment (bhargav265) [inline app/views/images/index.html.erb]: moving the logic to the view model is better

Comment (lei-appfolio) [reply app/views/images/index.html.erb]: OK, I will do this. Thank you!

File: test/view_models/images_index_view_test.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+require 'test_helper'
+
+class ImagesIndexViewTest < ActiveSupport::TestCase
+  def test_index_page_title
+    images_index_view = ImagesIndexView.new(search_tag: 'cute')
+    assert_equal 'All Images with tag [cute]', images_index_view.index_page_title
+
+    images_index_view = ImagesIndexView.new(search_tag: 'cat')
```
Comment (bhargav265) [inline test/view_models/images_index_view_test.rb]: Forgot to test the search tag blank case here

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -4,7 +4,10 @@ def new
   end
 
   def index
-    @images = Image.all
+    search_tag = search_tag_params[:search_tag]
+    @index_view_model = ImagesIndexView.new(search_tag: search_tag)
+    @images = search_tag.blank? ? Image.all : Image.tagged_with(search_tag)
```
Comment (bhargav265) [inline app/controllers/images_controller.rb]: I do not think you need this instance variable anymore

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -83,6 +86,91 @@ def test_index_image_order
...
+    url3 = 'https://img.apmcdn.org/5d9c531be5686d2572bcab206df39a230c44f642/uncropped/a85434-20161213-cat.jpg'
+    Image.create!(web_url: url1, mytag_list: 'cute, cat')
+    Image.create!(web_url: url2, mytag_list: 'cat, dog')
+    Image.create!(web_url: url3, mytag_list: 'cute, panda')
+
+    # case 1
+    get images_path(search_tag: '')
+    assert_response :ok
+
+    assert_select 'h1', 'All Images'
```
Comment (bhargav265) [inline test/controllers/images_controller_test.rb]: This is completely fine for now, we can stub the view model to return something we want and test it. This is especially used when the view model has very complex logic.

File: test/view_models/images_index_view_test.rb (hunk 2 above)
Comment (lei-appfolio) [reply test/view_models/images_index_view_test.rb]: OK!

File: app/controllers/images_controller.rb (hunk 3 above)
Comment (lei-appfolio) [reply app/controllers/images_controller.rb]: OK!

**Pull request merged**